### PR TITLE
Replace go-migrate with morph

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
Replaces `go-migrate` with morph.

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: store/sqlstore/store.go [hunk 1]
```diff
@@ -318,6 +319,14 @@ func (ss *SqlStore) initConnection() {
 		}
 	}
 
+	// When WithInstance is used in migration, the underlying driver connections are not tracked.
+	// So we will have to open a fresh connection for migrations and explicitly close it when all is done.
+	var err error
+	dataSource, err = ss.appendMultipleStatementsFlag(dataSource)
+	if err != nil {
+		panic(err)
```
Comment (agnivade) [inline store/sqlstore/store.go:327]: We should not use `multiStatements` for the whole application, just during migrations. This means that we are forced to open a new DB instance for MySQL. This is expected.

I'd suggest to bring this back inside `migrate` and check if the driver is MySQL. If yes, then call this and `setupConnection`. Else just reuse `ss.GetMasterX().DB.DB`

File: store/sqlstore/store.go [hunk 2]
```diff
@@ -1450,39 +1438,53 @@ func (ss *SqlStore) migrate(direction migrationDirection) error {
...
-	sourceDriver, err := bindata.WithInstance(source)
 	if err != nil {
 		return err
 	}
+	defer src.Close()
+
+	db := ss.GetMasterX().DB.DB
+	if db == nil {
+		return fmt.Errorf("db instance is nil")
+	}
```
Comment (agnivade) [inline store/sqlstore/store.go:1455]: Do we expect this to be nil, or it's just defence in depth?

File: scripts/mattermost-mysql-6.0.0.sql [hunk 3]
```diff
@@ -1177,15 +1177,15 @@ CREATE TABLE `Users` (
...
+-- Table structure for table `db_migrations`
 --
 
 /*!40101 SET @saved_cs_client     = @@character_set_client */;
 /*!40101 SET character_set_client = utf8 */;
-CREATE TABLE `schema_migrations` (
-  `version` bigint(20) NOT NULL,
-  `dirty` tinyint(1) NOT NULL,
-  PRIMARY KEY (`version`)
+CREATE TABLE `db_migrations` (
```
Comment (agnivade) [inline scripts/mattermost-mysql-6.0.0.sql:1185]: Is it correct to remove the old table? A 6.0 DB will already have `schema_migrations` table. Maybe we want to add the new table instead or change the schema of this table itself?

File: db/migrations/postgres/000046_create_users.down.sql [hunk 4]
```diff
@@ -1,5 +1,5 @@
 CREATE INDEX IF NOT EXISTS idx_users_email ON users (email);
 
-ALTER TABLE users DROP COLUMN IF NOT EXISTS remoteid;
+ALTER TABLE users DROP COLUMN IF EXISTS remoteid;
```
Comment (agnivade) [inline db/migrations/postgres/000046_create_users.down.sql:3]: Wow this was a bad bug :/

File: store/sqlstore/store.go (hunk 2 above)
Comment (isacikgoz) [reply store/sqlstore/store.go:1455]: Maybe it was be too defensive, it's actually going to exit app if there is a problem with opening the db. Can be removed.

File: scripts/mattermost-mysql-6.0.0.sql (hunk 3 above)
Comment (isacikgoz) [reply scripts/mattermost-mysql-6.0.0.sql:1185]: The structure of two tables are different, and also how they interpreted. I think we shouldn't try to use existing one. If the app is going to start with a fresh db it's not necessary to create a table for go-migrate.

File: db/migrations/postgres/000046_create_users.down.sql (hunk 4 above)
Comment (isacikgoz) [reply db/migrations/postgres/000046_create_users.down.sql:3]: Indeed it was, luckily morph caught these :)

**Review by agnivade — APPROVED** [review]
Fantastic work!

**Comment by isacikgoz** [comment]
Wanted to give a heads up that the new migration system can be merged into `master`. Before going to that path, I'd like to get your opinions and also answer any questions if you have any @spirosoik @stylianosrigas @gabrieljackson

**Comment by stylianosrigas** [comment]
@isacikgoz happy to see the morph coming into place. Do we expect any performance issues when we do the release?

**Comment by isacikgoz** [comment]
@stylianosrigas I'm not expecting any performance issues directly related by the implementation. But we need to be careful when setting timeout (`MigrationsStatementTimeoutSeconds`) for each migrations to run. Other than that, it should work w/o any issue. One of thing that we need to consider is that the tool is going to apply all migrations (72 files is going to be applied) in the first run because the migration table will be empty. There are a couple of indexes and table re-builds can happen. So I'd say we can either manually fill migration table to prevent migrations to run from scratch or pick a maintenance window and update.

**Comment by stylianosrigas** [comment]
Thanks @isacikgoz do you know in which release we are planning to include this? In the one next week or in the new year?

**Comment by isacikgoz** [comment]
I think it depends on how it would behave in the `master` branch, so far my tests didn't show any problem but I worked with fabricated data all the time. So testing it on the rc-test servers would be great before rolling out to the community. (I don't know if it aligns with our workflow tho 😅 ) But AFAIK we are not in a rush for making this released soon.

**Comment by agnivade** [comment]
> One of thing that we need to consider is that the tool is going to apply all migrations (72 files is going to be applied) in the first run because the migration table will be empty.

Considering all migrations are idempotent, it should just whizz through all of them right?

Also, to get some idea, we could get a schema dump from community and run this branch on it and compare with a schema dump later and see what gets changed. And then analyze and see if they make sense or not. And if so, then calculate how much time that can possibly take.

Would this make sense?

**Comment by isacikgoz** [comment]
There may be a few indexes added/removed along the way but majority will be no-op.

I think schema dump can help to see the impact, definitely makes sense.

**Comment by spirosoik** [comment]
@isacikgoz we plan to do it this weekend.

**Review by spirosoik — APPROVED** [review]

**Pull request merged**